src/utils/cosmosdb_model_versioned_ttl.ts

```diff
@@ -0,0 +1,207 @@
+    protected readonly partitionKey?: PartitionKey
+  ) {
+    super(
+      container,
+      newVersionedItemT,
+      retrievedItemT,
+      modelIdKey,
+      partitionKey
+    );
+  }
+
+  /**
+   * Using transactionalBatch updates the ttl field for all the versions
+   * of the document identified by the searchKey.
+   *
+   * @param searchKey
+   * @param ttl
+   * @returns Either a Cosmos Error or the number of updated elements
+   */
+  public updateTTLForAllVersions(
+    searchKey: DocumentSearchKey<TR, ModelIdKey, PartitionKey>,
+    ttl: RetrievedVersionedModelTTL["ttl"]
+  ): TE.TaskEither<CosmosErrors, number> {
+    return pipe(
+      this.findAllVersionsBySearchKey(searchKey),
+      TE.map(RA.rights),
+      TE.map(
+        RA.map(doc => ({
+          id: doc.id,
+          operationType: "Patch" as const,
+          resourceBody: {
+            operations: [
+              {
+                // `add` will create or update the path with the passed value
+                op: PatchOperationType.add,
+                path: `/ttl`,
+                value: ttl
+              }
+            ]
+          }
+        }))
+      ),
+      TE.map(RA.chunksOf(100)),
+      TE.map(
+        RA.map(chunk =>
+          pipe(chunk, RA.toArray, operations =>
+            this.batch(
+              operations,
+              `${
+                searchKey.length === 1 ? searchKey[0] : searchKey[1]
+              }` as NonEmptyString
+            )
+          )
+        )
+      ),
+      TE.chainW(TE.sequenceSeqArray),
+      TE.chainW(responses =>
+        pipe(
+          responses,
+          RA.filter(response => response.code !== 200),
+          TE.fromPredicate(
+            errors => errors.length === 0,
+            _ =>
+              CosmosErrorResponse({
+                message: `Error updating ttl for ${searchKey}`,
+                name: `Error updating ttl`
+              })
+          ),
+          TE.chainW(() =>
+            pipe(
+              responses,
+              RA.reduce(0, (v, r) => v + (r.result?.length ?? 0)),
+              TE.right
+            )
+          )
+        )
+      )
+    );
+  }
+
+  /**
+   * Given a searchKey returns all the version of a document
+   */
+
+  public findAllVersionsBySearchKey(
+    searchKey: DocumentSearchKey<TR, ModelIdKey, PartitionKey>
+  ): TE.TaskEither<CosmosErrors, ReadonlyArray<t.Validation<TR>>> {
+    const partitionKey = searchKey.length === 1 ? searchKey[0] : searchKey[1];
+    return pipe(
+      TE.tryCatch(
+        () =>
+          asyncIterableToArray(
+            this.getQueryIterator({
+              parameters: [
+                {
+                  name: "@partitionKey",
+                  value: partitionKey as JSONValue
+                }
+              ],
+              query: `SELECT * FROM m WHERE m.${
+                this.partitionKey ? this.partitionKey : this.modelIdKey
+              } = @partitionKey`
```

Comment:
To ensure that documents are processed on the right order, we need to sort by version
```suggestion
              query: `SELECT * FROM m WHERE m.${
                this.partitionKey ? this.partitionKey : this.modelIdKey
              } = @partitionKey ORDER BY m.version`
```

Reply:
In order not to weigh down the query I have preferred to order version at runtime (we know performances are even better).
Done in [c7d792a](https://github.com/pagopa/io-functions-commons/pull/296/commits/c7d792ad91044f66c0e186aad443e2721426a5b8)

Reply:
I was just checking to inform about possibile order by issue for not indexed properties. Remember the  infamous `order by id` for messages? 🤔

Reply:
That's why we didn't sort the query, but the results at runtime :D 
But please, remember that we are using ORER BY in `findLastVersionByModelId`